Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -253,3 +254,454 @@ theorem Language.IsContextFree.reverse (L : Language T) :
+          exact Option.noConfusion imposs
+        | tail _ hr =>
+          change r ∈ List.map _ g₁.rules ++ List.map _ g₂.rules at hr
+          rw [List.mem_append] at hr
+          cases hr with
+          | inl hr =>
+            rw [List.mem_map] at hr
+            exact hr
+          | inr hr =>
+            exfalso
+            rw [List.mem_map] at hr
+            rcases hr with ⟨_, -, rfl⟩
+            simp only [ContextFreeRule.lift, Function.comp_apply] at imposs
+            rw [Option.some_inj] at imposs
+            exact Sum.noConfusion imposs)⟩
+
+private def g₂g : LiftedContextFreeGrammar T :=
+  ⟨g₂, ContextFreeGrammar.union g₁ g₂, some ∘ Sum.inr, oN₂_of_N,
+    (fun x y hxy => Sum.inr_injective (Option.some_injective _ hxy)),
+    (by
+      intro x y hxy
+      cases x with
+      | none => right; rfl;
+      | some x₀ =>
+        cases y with
+        | none => right; exact hxy
+        | some y₀ =>
+          cases x₀ with
+          | inl =>
+            cases y₀ with
+            | inl =>
+              right
+              rfl
+            | inr =>
+              exfalso
+              simp [oN₂_of_N] at hxy
+          | inr =>
+            cases y₀ with
+            | inl =>
+              exfalso
+              simp [oN₂_of_N] at hxy
+            | inr =>
+              simp only [oN₂_of_N, Option.some.injEq] at hxy
+              left
+              rw [hxy]),
+    (fun _ => rfl),
+    (by
+      intro r _
+      apply List.mem_cons_of_mem
+      apply List.mem_cons_of_mem
+      apply List.mem_append_right
+      rw [List.mem_map]
+      use r),
+    (by
+      intro r ⟨hr, ⟨n₀, imposs⟩⟩
+      cases hr with
+      | head =>
+        exfalso
+        exact Option.noConfusion imposs
+      | tail _ hr =>
+        cases hr with
+        | head =>
+          exfalso
+          exact Option.noConfusion imposs
+        | tail _ hr =>
+          change r ∈ List.map _ g₁.rules ++ List.map _ g₂.rules at hr
+          rw [List.mem_append] at hr
+          cases hr with
+          | inl hr =>
+            exfalso
+            rw [List.mem_map] at hr
+            rcases hr with ⟨_, -, rfl⟩
+            simp only [ContextFreeRule.lift, Function.comp_apply] at imposs
+            rw [Option.some_inj] at imposs
+            exact Sum.noConfusion imposs
+          | inr hr =>
+            rw [List.mem_map] at hr
+            exact hr)⟩
+
+variable {w : List T}
+
+private lemma in_union_of_in_left (hw : w ∈ g₁.language) :
+    w ∈ (ContextFreeGrammar.union g₁ g₂).language := by
+  have deri_start :
+    (ContextFreeGrammar.union g₁ g₂).Derives [Symbol.nonterminal none]
+      [Symbol.nonterminal (some (Sum.inl g₁.initial))] := by
+    refine ContextFreeGrammar.Produces.single
+      ⟨⟨none, [Symbol.nonterminal (some (Sum.inl g₁.initial))]⟩, List.mem_cons_self .., ?_⟩
+    rw [ContextFreeRule.rewrites_iff]
+    use [], []
+    simp
+  exact deri_start.trans (Symbol.liftString_all_terminals g₁g.liftNT w ▸ g₁g.lift_derives hw)
```

Comment:
```suggestion
private lemma in_union_of_in_left (hw : w ∈ g₁.language) :
    w ∈ (ContextFreeGrammar.union g₁ g₂).language :=
  have deri_start :
    (ContextFreeGrammar.union g₁ g₂).Derives [Symbol.nonterminal none]
      [Symbol.nonterminal (some (Sum.inl g₁.initial))] := by
    refine ContextFreeGrammar.Produces.single
      ⟨⟨none, [Symbol.nonterminal (some (Sum.inl g₁.initial))]⟩, List.mem_cons_self .., ?_⟩
    rw [ContextFreeRule.rewrites_iff]
    use [], []
    simp
  deri_start.trans (Symbol.liftString_all_terminals g₁g.liftNT w ▸ g₁g.lift_derives hw)
```

Reply:
Don't you want to extract `deri_start` to a separate lemma?

Reply:
Good point, yes!

Reply:
You can try to prove `deri_start` only once and use it twice!

Reply:
When I extracted `deri_start`, lean complaint about a type mismatch. Is this due to some weird type inference?

Reply:
I'll fix it.

Reply:
Well, you didn't generalized it. You just copied the statement for `g₁.initial` which cannot work in the `g₂.language` lemma. I will require some deeper refactoring, which I don't have time for now.

Reply:
After discussing this with Martin on Zulip, we concluded that it is best to leave this minor duplication. The code needed to avoid it would be complex and not useful for anything else.